### PR TITLE
add wifi.on_connect and wifi.on_disconnect triggers

### PR DESCRIPTION
**Pull request by ChemicalXandco**
# What does this implement/fix?

add wifi.on_connect and wifi.on_disconnect triggers

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** https://github.com/esphome/esphome-docs/pull/2185

## Test Environment

- [ ] ESP32
- [ ] ESP32 IDF
- [x] ESP8266
- [x] RP2040

## Example entry for `config.yaml`:

```yaml
# Example config.yaml
wifi:
  on_connect:
    - logger.log: "connected"
  on_disconnect:
    - logger.log: "disconnected"
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Comment by nagyrobi** [comment]
I would suggest to implement this universally.
Not only applicable for wifi, but also for Ethernet component. Thus use `network.on_...` as names.

**Comment by ChemicalXandco** [comment]
> I would suggest to implement this universally. Not only applicable for wifi, but also for Ethernet component. Thus use `network.on_...` as names.

I think ethernet should be implemented separately and I cannot think of any advantages of unifying the triggers whilst it complicates the implementation slightly.

File: esphome/components/mdns/__init__.py [hunk 1]
```diff
@@ -29,7 +29,7 @@ def _remove_id_if_disabled(value):
 )
 
 
-@coroutine_with_priority(55.0)
+@coroutine_with_priority(-5.0)
```
Comment (RoboMagus) [inline esphome/components/mdns/__init__.py]: This doesn't  seem to be related to the connection triggers. Is it required?

File: esphome/components/wifi/wifi_component_esp32_arduino.cpp [hunk 2]
```diff
@@ -477,6 +477,7 @@ void WiFiComponent::wifi_event_callback_(esphome_wifi_event_id_t event, esphome_
       ESP_LOGV(TAG, "Event: Connected ssid='%s' bssid=" LOG_SECRET("%s") " channel=%u, authmode=%s", buf,
                format_mac_addr(it.bssid).c_str(), it.channel, get_auth_mode_str(it.authmode));
 
+      global_wifi_component->connect_trigger_->trigger();
```
Comment (RoboMagus) [inline esphome/components/wifi/wifi_component_esp32_arduino.cpp]: As this is part of the WiFiComponent that contains the triggers itself there should not be a reference to the `global_wifi_component`. Use `this` instead.

**Review by RoboMagus — CHANGES_REQUESTED** [review]
Would be a great addition, but I believe some minor rework is required.

File: esphome/components/mdns/__init__.py (hunk 1 above)
Comment (ChemicalXandco) [reply esphome/components/mdns/__init__.py]: When a trigger is added, somehow mdns `to_code` is called before the wifi `to_code` finishes despite having a lower priority which in turn inverts the import order of libraries ESP8266WiFi and ESP8266mDNS causing compilation to fail. This was the only way I could fix it without delving into the whole system of which I have no familiarity with unfortunately.

Comment (RoboMagus) [reply esphome/components/mdns/__init__.py]: What dev environment version are you on?

I tried to replicate using `devcontainer` within VSCode on `v2022.7.0-dev`. Without this change to the mdns __init__.py it compiles just fine.

Comment (ChemicalXandco) [reply esphome/components/mdns/__init__.py]: In my testing the problem ceases when certain components are enabled. Are you able to successfully compile the following?:
```yaml
esphome:
  name: example

esp8266:
  board: nodemcuv2

wifi:
  ssid: ExampleWifi
  password: ExamplePassword

  on_connect:
    - switch.turn_off: light_1

  on_disconnect:
    - switch.turn_on: light_1

switch:
  - platform: gpio
    id: light_1
    name: "Switch light 1"
    pin: D5
```

Comment (RoboMagus) [reply esphome/components/mdns/__init__.py]: Just gave it a try with the config you provided on a clean environment and it works just fine for me.
For the sake of curiosity, could you upload a full compilation log?

Comment (ChemicalXandco) [reply esphome/components/mdns/__init__.py]:
```
INFO Reading configuration test10.yaml...
INFO Generating C++ source...
INFO Core config or version changed, cleaning build files...
INFO Compiling app...
Processing example (board: nodemcuv2; framework: arduino; platform: platformio/espressif8266 @ 3.2.0)
-------------------------------------------------------------------------------------------------------------------------------------------------------
HARDWARE: ESP8266 80MHz, 80KB RAM, 4MB Flash
LDF: Library Dependency Finder -> https://bit.ly/configure-pio-ldf
Dependency Graph
|-- ESP8266mDNS @ 1.2
|-- ESP8266WiFi @ 1.0
Compiling .pioenvs/example/src/esphome/components/esp8266/core.cpp.o
Compiling .pioenvs/example/src/esphome/components/esp8266/gpio.cpp.o
Compiling .pioenvs/example/src/esphome/components/esp8266/preferences.cpp.o
Compiling .pioenvs/example/src/esphome/components/gpio/switch/gpio_switch.cpp.o
Compiling .pioenvs/example/src/esphome/components/mdns/mdns_component.cpp.o
Compiling .pioenvs/example/src/esphome/components/mdns/mdns_esp32_arduino.cpp.o
Compiling .pioenvs/example/src/esphome/components/mdns/mdns_esp8266.cpp.o
Compiling .pioenvs/example/src/esphome/components/mdns/mdns_esp_idf.cpp.o
Compiling .pioenvs/example/src/esphome/components/network/util.cpp.o
Compiling .pioenvs/example/src/esphome/components/switch/automation.cpp.o
Compiling .pioenvs/example/src/esphome/components/switch/switch.cpp.o
Compiling .pioenvs/example/src/esphome/components/wifi/wifi_component.cpp.o
Compiling .pioenvs/example/src/esphome/components/wifi/wifi_component_esp32_arduino.cpp.o
Compiling .pioenvs/example/src/esphome/components/wifi/wifi_component_esp8266.cpp.o
Compiling .pioenvs/example/src/esphome/components/wifi/wifi_component_esp_idf.cpp.o
Compiling .pioenvs/example/src/esphome/core/application.cpp.o
Compiling .pioenvs/example/src/esphome/core/color.cpp.o
Compiling .pioenvs/example/src/esphome/core/component.cpp.o
Compiling .pioenvs/example/src/esphome/core/component_iterator.cpp.o
Compiling .pioenvs/example/src/esphome/core/controller.cpp.o
Compiling .pioenvs/example/src/esphome/core/entity_base.cpp.o
Compiling .pioenvs/example/src/esphome/core/helpers.cpp.o
Compiling .pioenvs/example/src/esphome/core/log.cpp.o
Compiling .pioenvs/example/src/esphome/core/scheduler.cpp.o
Compiling .pioenvs/example/src/esphome/core/util.cpp.o
Compiling .pioenvs/example/src/main.cpp.o
Generating LD script .pioenvs/example/ld/local.eagle.app.v6.common.ld
Compiling .pioenvs/example/lib565/ESP8266mDNS/ESP8266mDNS.cpp.o
In file included from /home/[user]/.platformio/packages/framework-arduinoespressif8266/libraries/ESP8266mDNS/src/ESP8266mDNS.h:48,
                 from /home/[user]/.platformio/packages/framework-arduinoespressif8266/libraries/ESP8266mDNS/src/ESP8266mDNS.cpp:23:
/home/[user]/.platformio/packages/framework-arduinoespressif8266/libraries/ESP8266mDNS/src/LEAmDNS.h:106:10: fatal error: WiFiUdp.h: No such file or directory

*****************************************************************
* Looking for WiFiUdp.h dependency? Check our library registry!
*
* CLI  > platformio lib search "header:WiFiUdp.h"
* Web  > https://registry.platformio.org/search?q=header:WiFiUdp.h
*
*****************************************************************

  106 | #include "WiFiUdp.h"
      |          ^~~~~~~~~~~
compilation terminated.
*** [.pioenvs/example/lib565/ESP8266mDNS/ESP8266mDNS.cpp.o] Error 1
============================================================= [FAILED] Took 10.69 seconds =============================================================
```

> Just gave it a try with the config you provided on a clean environment and it works just fine for me.

In that case I shall omit this change and investigate a better solution.

Comment (RoboMagus) [reply esphome/components/mdns/__init__.py]: I just discovered the option to have increased verbosity when compiling from within esphome. 
`esphome -v compile test10.yaml`

Could you give that a try too? That would be useful when creating an issue for it.

**Review by RoboMagus — APPROVED** [review]

**Comment by anatoly-savchenkov** [comment]
I was using these changes for some time on my dev board and found that they break safe mode. Generated code without these changes:
```
  wifi_wificomponent = new wifi::WiFiComponent();
  wifi_wificomponent->set_use_address("lytko101.local");
  wifi::WiFiAP wifi_wifiap = wifi::WiFiAP();
  wifi_wificomponent->set_ap(wifi_wifiap);
  wifi_wificomponent->set_ap_timeout(60000);
  wifi_wificomponent->set_reboot_timeout(900000);
  wifi_wificomponent->set_power_save_mode(wifi::WIFI_POWER_SAVE_NONE);
  wifi_wificomponent->set_fast_connect(false);
  wifi_wificomponent->set_output_power(20.0f);
  wifi_wificomponent->set_component_source("wifi");
  App.register_component(wifi_wificomponent); <-- THIS LINE IS ESSENTIAL
...
  ota_otacomponent = new ota::OTAComponent();
  ota_otacomponent->set_port(8266);
  ota_otacomponent->set_component_source("ota");
  App.register_component(ota_otacomponent);
  if (ota_otacomponent->should_enter_safe_mode(10, 300000)) return; <-- This line terminates setup() function leaving only WiFi, mDNS and OTA enabled
  ```

Code with these new triggers looks like the following:
```
  wifi_wificomponent = new wifi::WiFiComponent();
  wifi_wificomponent->set_use_address("lytko101.local");
  wifi::WiFiAP wifi_wifiap = wifi::WiFiAP();
  wifi_wificomponent->set_ap(wifi_wifiap);
  wifi_wificomponent->set_ap_timeout(60000);
  wifi_wificomponent->set_reboot_timeout(900000);
  wifi_wificomponent->set_power_save_mode(wifi::WIFI_POWER_SAVE_NONE);
  wifi_wificomponent->set_fast_connect(false);
  wifi_wificomponent->set_output_power(20.0f);
  automation_8 = new Automation<>(wifi_wificomponent->get_connect_trigger());
....

  ota_otacomponent = new ota::OTAComponent();
  ota_otacomponent->set_port(8266);
  ota_otacomponent->set_component_source("ota");
  App.register_component(ota_otacomponent);
  if (ota_otacomponent->should_enter_safe_mode(10, 300000)) return; <-- WE STOP HERE
...
  automation_9->add_actions({lambdaaction_8});
  wifi_wificomponent->set_component_source("wifi");
  App.register_component(wifi_wificomponent); <--THIS LINE IS NOT REACHED IN CASE OF SAFE OTA MODE, WIFI IS NOT STARTED
```
So, I believe something is not right with priorities.

**Comment by ChemicalXandco** [comment]
@anatoly-savchenkov does the latest commit fix it?
thanks

**Comment by anatoly-savchenkov** [comment]
Thanks, @ChemicalXandco! Generated code looks good to me now. Later during the weekend I'll be able to test the changes as well.

**Comment by anatoly-savchenkov** [comment]
I confirm, safe mode works ok now. I would really like to see this PR merged.

**Comment by anatoly-savchenkov** [comment]
Another question to discuss. Imagine wifi STA connection is unsuccessful after boot and fallback AP is brought up. What should happen to these triggers? Nothing, call on_disconnect or even call on_connect. Should we have another trigger(s) for AP states?

**Comment by anatoly-savchenkov** [comment]
This is not stale. It's a useful change and it needs a review.

**Comment by anatoly-savchenkov** [comment]
Not stale and still waits for a review

File: esphome/components/wifi/__init__.py [hunk 3]
```diff
@@ -392,9 +396,16 @@ async def to_code(config):
...
+    # Register before OTA safe mode check
     await cg.register_component(var, config)
 
+    for conf in config.get(CONF_ON_CONNECT, []):
+        # OTA safe mode check invoked through here
+        await automation.build_automation(var.get_connect_trigger(), [], conf)
+
+    for conf in config.get(CONF_ON_DISCONNECT, []):
+        await automation.build_automation(var.get_disconnect_trigger(), [], conf)
+
```
Comment (oxan) [inline esphome/components/wifi/__init__.py:448]: That comment was a leftover from a long, long time ago when the ota component was part of the wifi component. There's nothing here that actually marks the OTA safe mode check, and doing it like this does in fact end up with the triggers before the safe mode check (which is not desirable).

There should be some sort of synchronization with the OTA component to put the check in the correct place, though I'm not immediately sure how that would look.

File: esphome/components/wifi/wifi_component_esp32_arduino.cpp [hunk 4]
```diff
@@ -477,6 +477,7 @@ void WiFiComponent::wifi_event_callback_(esphome_wifi_event_id_t event, esphome_
       ESP_LOGV(TAG, "Event: Connected ssid='%s' bssid=" LOG_SECRET("%s") " channel=%u, authmode=%s", buf,
                format_mac_addr(it.bssid).c_str(), it.channel, get_auth_mode_str(it.authmode));
 
+      this->connect_trigger_->trigger();
```
Comment (oxan) [inline esphome/components/wifi/wifi_component_esp32_arduino.cpp]: I think this doesn't run from the loop task, but instead from whatever task Arduino uses to callback to us. As ESPHome isn't made to be concurrency-safe, that can cause serious issues if we let arbitrary user code run.

**Review by oxan — DISMISSED** [review]

**Comment by nagyrobi** [comment]
Now that we have manual wifi.disable and wifi.enable [actions](https://esphome.io/components/wifi.html#wifi-disable-action), this PR is worth it.

**Comment by anatoly-savchenkov** [comment]
@jesserockz, any chance to have this merged?

**Review by kbx81 — APPROVED** [review]
LGTM, thanks!

**Pull request merged**